core/src/Kokkos_HostSpace.hpp

```diff
@@ -202,10 +202,17 @@ struct DeepCopy<HostSpace, HostSpace, ExecutionSpace> {
   }
 
   DeepCopy(const ExecutionSpace& exec, void* dst, const void* src, size_t n) {
-    exec.fence(
-        "Kokkos::Impl::DeepCopy<HostSpace, HostSpace, "
-        "ExecutionSpace>::DeepCopy: fence before copy");
-    hostspace_parallel_deepcopy_async(dst, src, n);
+    // If ExecutionSpace is a Device ExecutionSpace, we should call
+    // this from DefaultHostExecutionSpace
+    if constexpr (std::is_same_v<ExecutionSpace, DefaultExecutionSpace> &&
+                  !std::is_same_v<ExecutionSpace, DefaultHostExecutionSpace>) {
+      exec.fence(
+          "Kokkos::Impl::DeepCopy<HostSpace, HostSpace, "
+          "ExecutionSpace>::DeepCopy: fence before copy");
+      hostspace_parallel_deepcopy_async(dst, src, n);
```

Comment:
Uhoh I think semantically we need to do actually something more complicated than this. To be fully correct in this case I believe we need prevent any operations on exec to overlap with the deep_copy. Which means we likely need to 
- acquire exec lock
- wait for all exec operations to finish
- perform the deep_copy (and wait for it to finish)
- release the exec lock

Reply:
That is indeed more complicated.
Could you please detail why `exec.fence()` is not good enough? 
Also, is there a way to acquire exec lock for all execution spaces?
What I have in mind is something like this, which should not work
```C++
auto *internal_instance = exec.impl_internal_space_instance();
// I guess this is not available for device execution spaces
std::lock_guard<std::mutex> lock(internal_instance->m_instance_mutex);
internal_instance->fence(
    "Kokkos::Impl::DeepCopy<HostSpace, HostSpace, "
    "ExecutionSpace>::DeepCopy: fence before copy");
// How to wait for this to finish?
hostspace_parallel_deepcopy_async(dst, src, n);
```

Reply:
Essentially the problem is that the next kernel submitted to exec could overlap with the async host deep_copy, but semantically the deep_copy needs to be sequenced in the device queue. If you only have a single process thread, then the overlap could happen if host exec is async (i.e. HPX) which we can fix by fencing after the host deep_copy the host exec. If there are multiple (non-kokkos) threads, the overlap could come from another thread submitting to exec while the deep copy is running. But I am ok fixing that separately.

Reply:
We only use this branch for a device execution space instance which isn't able to access the memory in question anyway. The only problem could arise if we have another host execution space (different from the one used here) accesses the memory from a different thread but we don't promise anything in that case anyway. If we happen to use the same execution space instance, we enqueue correctly.